<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>EuroBioc2020 NewWave workshop • SurfingNewWave</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="EuroBioc2020 NewWave workshop">
<meta property="og:description" content="SurfingNewWave">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SurfingNewWave</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette.html">EuroBioc2020 NewWave workshop</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/fedeago/SurfingNewWave">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>EuroBioc2020 NewWave workshop</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/fedeago/SurfingNewWave/blob/master/vignettes/vignette.Rmd"><code>vignettes/vignette.Rmd</code></a></small>
      <div class="hidden name"><code>vignette.Rmd</code></div>

    </div>

    
    
<div id="introduction-to-newwave" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-to-newwave" class="anchor"></a>Introduction to NewWave</h1>
<p>EuroBioc 2020: 14-18 December</p>
<div id="workshop-description" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h2>
<p>The fast development of single cell sequencing technologies in the recent years has generated a gap between the throughput of the experiments and the capability of analizing these generated data. In this package, we implement mini-batch stochastic gradient descent and the possibility to work with HDF5 files. We decided to use a negative binomial model following the observation that droplet sequencing technologies do not induce zero inflation in the data. Thanks to these improvements and the possibility of massively parallelize the estimation process using PSOCK clusters, we are able to speed up the computation.</p>
<div id="r-bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<ul>
<li>SharedObject</li>
<li>DelayedArray</li>
<li>SingleCellExperiment</li>
<li>splatter</li>
<li>TENxBrainData</li>
</ul>
</div>
</div>
<div id="workshop-goals-and-objectives" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h2>
<p>In this workshop I am going to show how to use NewWave with different object and how to choose the optimal parameter related to your hardware.</p>
</div>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>NewWave assumes a negative binomial distribution where the mean parameter is related to each single observation while the dispersion parameter can be equal for all observation or related to each single gene.</p>
<p><span class="math display">\[Y_{ij} \sim NB(\mu_{ij},\theta_j)\]</span> And this is how the parameters are estimated:</p>
<p><span class="math display">\[\ln(\mu_{ij}) = \Big(X\beta + (V\gamma)^T + W\alpha \Big)_{ij} \]</span> <span class="math display">\[\ln(\theta_j)   = \zeta_j\]</span> These are the steps used for estimate parameters : * Initialization of regression parameters parameters using ridge regression, and low-rank rapresentation with PCA. * Optimization of dispersion parameters, if genewise massively parallelizable. * Optimization of the cell-related parameters, using MLE, massively parallelizable. * Optimization of the gene-related parameters, using MLE, massive parallelizable.</p>
<p>While the initialization is done only once at the beginnig of the process the other three steps are done iteratively till convergence.</p>
<p>In order to reduce the memory consumption it uses a PSOCK cluster combined with the R package SharedObject that allow to share a matrix between different cores avoiding memory duplication. Thanks to that we can massively parallelize the estimation process with huge benefit in terms of time consumption. We can reduce even more the time consumption using some mini batch approaches on the different steps of the optimization.</p>
<p>I am going to show how to use NewWave with example data generated with Splatter.</p>
<p>Splatter is a R package designed to generate single cell transcriptomic data with a high level of customization.</p>
<p>In our implementation we want a batch effect to be corrected, the additionals batch parameters are setted in order to create more complex batch effect.</p>
<p>You can think to the group as a cell type, we will use those group to test the resulted low dimensional representation through a cluster analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span>
  <span class="op">{</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Oshlack/splatter">splatter</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">irlba</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jkrijthe/Rtsne">Rtsne</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://mclust-org.github.io/mclust/">mclust</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">NewWave</span><span class="op">)</span>
  <span class="op">}</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/splatter/man/newParams.html">newSplatParams</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">N</span><span class="op">=</span><span class="fl">1000</span>
<span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/splatter/man/splatSimulate.html">splatSimulateGroups</a></span><span class="op">(</span><span class="va">params</span>,batchCells<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">N</span><span class="op">/</span><span class="fl">2</span>,<span class="va">N</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span>,
                           group.prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0.1</span>,<span class="fl">10</span><span class="op">)</span>,
                           de.prob <span class="op">=</span> <span class="fl">0.2</span>,
                           batch.facLoc <span class="op">=</span> <span class="fl">2</span>,
                           batch.facScale <span class="op">=</span> <span class="fl">1</span>,
                           verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </code></pre></div>
<p>The resault of splatter is a SingleCellExperiment with the generated data in the assays counts.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 10000 1000 
## metadata(1): Params
## assays(6): BatchCellMeans BaseCellMeans ... TrueCounts counts
## rownames(10000): Gene1 Gene2 ... Gene9999 Gene10000
## rowData names(16): Gene BaseGeneMean ... DEFacGroup9 DEFacGroup10
## colnames(1000): Cell1 Cell2 ... Cell999 Cell1000
## colData names(4): Cell Batch Group ExpLibSize
## reducedDimNames(0):
## altExpNames(0):</code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "matrix" "array"</code></pre>
<p>As you can see we have the number of cell that we choose(N=500) an 10 000 genes. For this workshop we will choose only the 1000 most variable genes.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hvg</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">hvg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span>
<span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">hvg</span>,decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span>,<span class="op">]</span></code></pre></div>
<p>In a SingleCellExperiment the colData slot contains the sample level metadata. In our situation there are two important variable, Batch and Group.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">data</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 1000 rows and 4 columns
##                 Cell       Batch    Group ExpLibSize
##          &lt;character&gt; &lt;character&gt; &lt;factor&gt;  &lt;numeric&gt;
## Cell1          Cell1      Batch1  Group8     60315.6
## Cell2          Cell2      Batch1  Group1     72936.1
## Cell3          Cell3      Batch1  Group10    61558.7
## Cell4          Cell4      Batch1  Group7     69118.2
## Cell5          Cell5      Batch1  Group6     46349.5
## ...              ...         ...      ...        ...
## Cell996      Cell996      Batch2  Group3     61310.1
## Cell997      Cell997      Batch2  Group10    59189.1
## Cell998      Cell998      Batch2  Group2     67265.0
## Cell999      Cell999      Batch2  Group10    55379.7
## Cell1000    Cell1000      Batch2  Group3     54241.7</code></pre>
<p><strong>IMPORTANT:</strong> For batch effecr removal the batch variable must be a factor</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data</span><span class="op">$</span><span class="va">Batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">Batch</span><span class="op">)</span></code></pre></div>
<p>NewWave takes as input raw data, not normalized. So now we have all the object needed to start.</p>
<p>We can see the how the cells are distributed between group and batch.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/irlba/man/prcomp_irlba.html">prcomp_irlba</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span>,n<span class="op">=</span><span class="fl">10</span><span class="op">)</span>
<span class="va">plot_data</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html">Rtsne</a></span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">x</span><span class="op">)</span><span class="op">$</span><span class="va">Y</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot_data</span><span class="op">$</span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">Batch</span>
<span class="va">plot_data</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">Group</span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">plot_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">X1</span>,y<span class="op">=</span><span class="va">X2</span>,col<span class="op">=</span><span class="va">group</span>, shape<span class="op">=</span><span class="va">batch</span><span class="op">)</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>Just for a later comparison we can see the performance of clustering on PCA low dimensional representation.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">x</span>, <span class="fl">10</span><span class="op">)</span>

<span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html">adjustedRandIndex</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">$</span><span class="va">cluster</span>, <span class="va">data</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0.04704868</code></pre>
<p>There is a clear batch effect between the cells.</p>
<p>Let’s try to correct it.</p>
</div>
<div id="newwave" class="section level1">
<h1 class="hasAnchor">
<a href="#newwave" class="anchor"></a>NewWave</h1>
<p>I am going to show different implementation and the suggested way to use them with the given hardware.</p>
<p>Some basic advise:</p>
<ul>
<li>Verbose option has default FALSE, in this vignette I will change it for explanatory intentions, don’t do it with big dataset because it can sensibly slower the computation</li>
<li>There are no concern about the dimension of mini-batches, I always used the 10% of the observations</li>
</ul>
<div id="standard-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#standard-usage" class="anchor"></a>Standard usage</h2>
<p>I am going to present the stadard implementation of NewWave, it uses 1 cores, it estimates a dispersion parameters equal for all observation and uses all the observation in the estimation process.</p>
<p>The K parameter represents the number of dimension of the latent rapresentation.</p>
<p>The X parameter is used to indicates which variable represent the batch effect, in the same way can be inserted other cell-related variable and if you need some gene related variable those can be inserted in V.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NewWave/man/newWave.html">newWave</a></span><span class="op">(</span><span class="va">data</span>, X <span class="op">=</span> <span class="st">"~Batch"</span>, K<span class="op">=</span><span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## Time of setup
##    user  system elapsed 
##   0.008   0.005   0.272 
## Time of initialization
##    user  system elapsed 
##   0.046   0.027   0.822</code></pre>
<pre><code>## Iteration 1</code></pre>
<pre><code>## penalized log-likelihood = -2039588.42312985</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   1.193   0.072   1.236</code></pre>
<pre><code>## after optimize dispersion = -1761061.82409032</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##    0.00    0.00    9.86</code></pre>
<pre><code>## after right optimization= -1713687.47900254</code></pre>
<pre><code>## after orthogonalization = -1713683.75718161</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.030   0.016   9.457</code></pre>
<pre><code>## after left optimization= -1707523.64825206</code></pre>
<pre><code>## after orthogonalization = -1707523.27582134</code></pre>
<pre><code>## Iteration 2</code></pre>
<pre><code>## penalized log-likelihood = -1707523.27582134</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   1.307   0.112   1.289</code></pre>
<pre><code>## after optimize dispersion = -1706425.20511649</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##   0.000   0.000  10.104</code></pre>
<pre><code>## after right optimization= -1705900.36994457</code></pre>
<pre><code>## after orthogonalization = -1705900.30326209</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.016   0.032  10.107</code></pre>
<pre><code>## after left optimization= -1705684.63698912</code></pre>
<pre><code>## after orthogonalization = -1705684.60936773</code></pre>
<pre><code>## Iteration 3</code></pre>
<pre><code>## penalized log-likelihood = -1705684.60936773</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   1.295   0.072   1.236</code></pre>
<pre><code>## after optimize dispersion = -1705681.14456135</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##   0.001   0.000   9.393</code></pre>
<pre><code>## after right optimization= -1705571.1538119</code></pre>
<pre><code>## after orthogonalization = -1705571.14479676</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.022   0.024   9.717</code></pre>
<pre><code>## after left optimization= -1705504.25290166</code></pre>
<pre><code>## after orthogonalization = -1705504.25143127</code></pre>
<pre><code>## Iteration 4</code></pre>
<pre><code>## penalized log-likelihood = -1705504.25143127</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   1.301   0.068   1.238</code></pre>
<pre><code>## after optimize dispersion = -1705504.12775755</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##   0.000   0.001   8.131</code></pre>
<pre><code>## after right optimization= -1705457.08298441</code></pre>
<pre><code>## after orthogonalization = -1705457.0788763</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.026   0.024   9.166</code></pre>
<pre><code>## after left optimization= -1705423.93217403</code></pre>
<pre><code>## after orthogonalization = -1705423.93133366</code></pre>
<p>In order to make it faster you can increase the number of cores using <code>children</code> parameter:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NewWave/man/newWave.html">newWave</a></span><span class="op">(</span><span class="va">data</span>,X <span class="op">=</span> <span class="st">"~Batch"</span>, K<span class="op">=</span><span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>, children<span class="op">=</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
<pre><code>## Time of setup
##    user  system elapsed 
##   0.011   0.000   0.283 
## Time of initialization
##    user  system elapsed 
##   0.049   0.008   0.675</code></pre>
<pre><code>## Iteration 1</code></pre>
<pre><code>## penalized log-likelihood = -2039588.4231303</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   1.173   0.028   1.153</code></pre>
<pre><code>## after optimize dispersion = -1761061.82409008</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##   0.001   0.000  10.923</code></pre>
<pre><code>## after right optimization= -1713687.47899673</code></pre>
<pre><code>## after orthogonalization = -1713683.7571758</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.028   0.012   4.874</code></pre>
<pre><code>## after left optimization= -1707523.64824653</code></pre>
<pre><code>## after orthogonalization = -1707523.27581579</code></pre>
<pre><code>## Iteration 2</code></pre>
<pre><code>## penalized log-likelihood = -1707523.27581579</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   1.320   0.056   1.246</code></pre>
<pre><code>## after optimize dispersion = -1706425.20510817</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##   0.001   0.000  10.753</code></pre>
<pre><code>## after right optimization= -1705900.36993789</code></pre>
<pre><code>## after orthogonalization = -1705900.30325539</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.029   0.020   5.199</code></pre>
<pre><code>## after left optimization= -1705684.63517902</code></pre>
<pre><code>## after orthogonalization = -1705684.60756164</code></pre>
<pre><code>## Iteration 3</code></pre>
<pre><code>## penalized log-likelihood = -1705684.60756164</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   1.330   0.055   1.254</code></pre>
<pre><code>## after optimize dispersion = -1705681.14284763</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##    0.00    0.00    9.32</code></pre>
<pre><code>## after right optimization= -1705571.1695862</code></pre>
<pre><code>## after orthogonalization = -1705571.1605932</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.011   0.013   5.013</code></pre>
<pre><code>## after left optimization= -1705504.20779332</code></pre>
<pre><code>## after orthogonalization = -1705504.20630812</code></pre>
<pre><code>## Iteration 4</code></pre>
<pre><code>## penalized log-likelihood = -1705504.20630812</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   1.319   0.068   1.255</code></pre>
<pre><code>## after optimize dispersion = -1705504.08277459</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##   0.001   0.000   9.065</code></pre>
<pre><code>## after right optimization= -1705457.02800826</code></pre>
<pre><code>## after orthogonalization = -1705457.02385328</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.016   0.032   4.922</code></pre>
<pre><code>## after left optimization= -1705423.80432659</code></pre>
<pre><code>## after orthogonalization = -1705423.80345976</code></pre>
</div>
<div id="commonwise-dispersion-and-minibatch-approaches" class="section level2">
<h2 class="hasAnchor">
<a href="#commonwise-dispersion-and-minibatch-approaches" class="anchor"></a>Commonwise dispersion and minibatch approaches</h2>
<p>If you do not have an high number of cores to run newWave this is the fastest way to run.</p>
<p>Each of these three steps can be accelerated using mini batch, the number of observation is settled with these parameters:</p>
<ul>
<li>n_gene_disp : Number of genes to use in the dispersion optimization</li>
<li>n_cell_par : Number of cells to use in the cells related parameters optimization</li>
<li>n_gene_par : Number of genes to use in the genes related parameters optimization</li>
</ul>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NewWave/man/newWave.html">newWave</a></span><span class="op">(</span><span class="va">data</span>,X <span class="op">=</span> <span class="st">"~Batch"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>,K<span class="op">=</span><span class="fl">10</span>, children<span class="op">=</span><span class="fl">2</span>,
                n_gene_disp <span class="op">=</span> <span class="fl">100</span>, n_gene_par <span class="op">=</span> <span class="fl">100</span>, n_cell_par <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></code></pre></div>
<pre><code>## Time of setup
##    user  system elapsed 
##   0.011   0.000   0.284 
## Time of initialization
##    user  system elapsed 
##   0.043   0.024   0.651</code></pre>
<pre><code>## Iteration 1</code></pre>
<pre><code>## penalized log-likelihood = -2039588.42312966</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   1.180   0.028   1.163</code></pre>
<pre><code>## after optimize dispersion = -1761061.82409047</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##   0.000   0.000  10.365</code></pre>
<pre><code>## after right optimization= -1713687.47900281</code></pre>
<pre><code>## after orthogonalization = -1713683.75718187</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.017   0.008   4.862</code></pre>
<pre><code>## after left optimization= -1707523.64825239</code></pre>
<pre><code>## after orthogonalization = -1707523.27582166</code></pre>
<pre><code>## Iteration 2</code></pre>
<pre><code>## penalized log-likelihood = -1707523.27582166</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   0.438   0.096   0.402</code></pre>
<pre><code>## after optimize dispersion = -1706425.61873959</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##   0.001   0.000   2.130</code></pre>
<pre><code>## after right optimization= -1706314.13185101</code></pre>
<pre><code>## after orthogonalization = -1706314.12339951</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.020   0.016   0.552</code></pre>
<pre><code>## after left optimization= -1706307.64922279</code></pre>
<pre><code>## after orthogonalization = -1706307.64894363</code></pre>
<pre><code>## Iteration 3</code></pre>
<pre><code>## penalized log-likelihood = -1706307.64894363</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   0.473   0.060   0.402</code></pre>
<pre><code>## after optimize dispersion = -1706307.64894363</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##   0.000   0.000   1.869</code></pre>
<pre><code>## after right optimization= -1706225.83894611</code></pre>
<pre><code>## after orthogonalization = -1706225.83305468</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.012   0.036   0.535</code></pre>
<pre><code>## after left optimization= -1706215.59998116</code></pre>
<pre><code>## after orthogonalization = -1706215.59943099</code></pre>
</div>
<div id="genewise-dispersion-mini-batch" class="section level2">
<h2 class="hasAnchor">
<a href="#genewise-dispersion-mini-batch" class="anchor"></a>Genewise dispersion mini-batch</h2>
<p>If you have a lot of core disposable or you want to estimate a genewise dispersion parameter this is the fastest configuration:</p>
<p><strong>IMPORTANT</strong>:do not use n_gene_disp in this case, it will slower the computation.</p>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NewWave/man/newWave.html">newWave</a></span><span class="op">(</span><span class="va">data</span>,X <span class="op">=</span> <span class="st">"~Batch"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>,K<span class="op">=</span><span class="fl">10</span>, children<span class="op">=</span><span class="fl">2</span>,
                n_gene_par <span class="op">=</span> <span class="fl">100</span>, n_cell_par <span class="op">=</span> <span class="fl">100</span>, commondispersion <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code>## Time of setup
##    user  system elapsed 
##   0.011   0.000   0.278 
## Time of initialization
##    user  system elapsed 
##   0.053   0.004   0.641</code></pre>
<pre><code>## Iteration 1</code></pre>
<pre><code>## penalized log-likelihood = -2039588.42312965</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   1.177   0.024   1.153</code></pre>
<pre><code>## after optimize dispersion = -1761061.82409002</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##   0.000   0.000  10.277</code></pre>
<pre><code>## after right optimization= -1713687.4790026</code></pre>
<pre><code>## after orthogonalization = -1713683.75718167</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.016   0.017   4.878</code></pre>
<pre><code>## after left optimization= -1707523.64825382</code></pre>
<pre><code>## after orthogonalization = -1707523.27582308</code></pre>
<pre><code>## Iteration 2</code></pre>
<pre><code>## penalized log-likelihood = -1707523.27582308</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   0.133   0.068   0.978</code></pre>
<pre><code>## after optimize dispersion = -1699896.68204854</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##   0.001   0.000   2.453</code></pre>
<pre><code>## after right optimization= -1699807.53377703</code></pre>
<pre><code>## after orthogonalization = -1699807.51949525</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.013   0.028   0.583</code></pre>
<pre><code>## after left optimization= -1699759.85435883</code></pre>
<pre><code>## after orthogonalization = -1699759.85382331</code></pre>
<pre><code>## Iteration 3</code></pre>
<pre><code>## penalized log-likelihood = -1699759.85382331</code></pre>
<pre><code>## Time of dispersion optimization
##    user  system elapsed 
##   0.156   0.048   0.459</code></pre>
<pre><code>## after optimize dispersion = -1699760.1360841</code></pre>
<pre><code>## Time of right optimization
##    user  system elapsed 
##   0.001   0.000   2.147</code></pre>
<pre><code>## after right optimization= -1699687.52386648</code></pre>
<pre><code>## after orthogonalization = -1699687.51950177</code></pre>
<pre><code>## Time of left optimization
##    user  system elapsed 
##   0.010   0.009   0.588</code></pre>
<pre><code>## after left optimization= -1699642.11626436</code></pre>
<pre><code>## after orthogonalization = -1699642.11572079</code></pre>
<p>Now I can use the latent dimension rapresentation for visualization purpose:</p>
<div class="sourceCode" id="cb164"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">latent</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>

<span class="va">tsne_latent</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html">Rtsne</a></span><span class="op">(</span><span class="va">latent</span><span class="op">)</span><span class="op">$</span><span class="va">Y</span><span class="op">)</span>
<span class="va">tsne_latent</span><span class="op">$</span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">Batch</span>
<span class="va">tsne_latent</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">Group</span></code></pre></div>
<div class="sourceCode" id="cb165"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">tsne_latent</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">X1</span>,y<span class="op">=</span><span class="va">X2</span>,col<span class="op">=</span><span class="va">group</span>, shape<span class="op">=</span><span class="va">batch</span><span class="op">)</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>or for clustering:</p>
<div class="sourceCode" id="cb166"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span><span class="op">(</span><span class="va">latent</span>, <span class="fl">10</span><span class="op">)</span>

<span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html">adjustedRandIndex</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">$</span><span class="va">cluster</span>, <span class="va">data</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 1</code></pre>
</div>
</div>
<div id="newwave-on-delayedarray" class="section level1">
<h1 class="hasAnchor">
<a href="#newwave-on-delayedarray" class="anchor"></a>NewWave on DelayedArray</h1>
<p>In this package we can find a huge single cell dataset made by 1.3 M cell and 27 K genes.</p>
<p>Thanks to the hdf5 format that we can use trougth the DelayedArray framework this dataset is not on our RAM but it can be read and used anyway.</p>
<div class="sourceCode" id="cb168"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TENxBrainData</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb169"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tenx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxBrainData/man/TENxBrainData.html">TENxBrainData</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## using temporary cache /tmp/RtmplHhfqw/BiocFileCache</code></pre>
<pre><code>## snapshotDate(): 2020-12-14</code></pre>
<pre><code>## see ?TENxBrainData and browseVignettes('TENxBrainData') for documentation</code></pre>
<pre><code>## downloading 1 resources</code></pre>
<pre><code>## retrieving 1 resource</code></pre>
<pre><code>## loading from cache</code></pre>
<p>Let’s see how is made this SingleCellExperiment</p>
<div class="sourceCode" id="cb176"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tenx</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 27998 1306127 
## metadata(0):
## assays(1): counts
## rownames: NULL
## rowData names(2): Ensembl Symbol
## colnames(1306127): AAACCTGAGATAGGAG-1 AAACCTGAGCGGCTTC-1 ...
##   TTTGTCAGTTAAAGTG-133 TTTGTCATCTGAAAGA-133
## colData names(4): Barcode Sequence Library Mouse
## reducedDimNames(0):
## altExpNames(0):</code></pre>
<p>It has the slot counts were the date are stored exaclty how is needed by NewWave but in this case we do not have the batch effect and so we will only reduce the dimensionality of the dataset.</p>
<p>Despite of the huge dimension of the dataset it’s memory size is really small:</p>
<div class="sourceCode" id="cb178"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html">object.size</a></span><span class="op">(</span><span class="va">tenx</span><span class="op">)</span></code></pre></div>
<pre><code>## 304126984 bytes</code></pre>
<p>only 304 MB.</p>
<p>This is possible because the counts slot is not a matrix.</p>
<div class="sourceCode" id="cb180"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span><span class="op">(</span><span class="va">tenx</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "DelayedMatrix"
## attr(,"package")
## [1] "DelayedArray"</code></pre>
<p>In order to manage this type of file NewWave uses a set of proper method with some strength and some weakness.</p>
<p>I am not going to show the exact implementation but I will explain how to use NewWave in this situation and why.</p>
<p>The biggest strength of using DelayedArray and HDF5 files is the reduction of the RAM consumption, the cost of that is that we read only a small piece, called chunck, of the dataset every time and so if it is done not properly in can strongly slower the computation expecially when it is applied to an iterative optimization such as the one used by NewWave.</p>
<p>For this reason I strongly discourage to set <code>verbose = T</code>, this will force NewWave to read the dataset, chunk by chunk, several time for each iteration.</p>
<p>In this first version of the package this implementation is really effective only when applied on big dataset, let’s say at least 30K cell.</p>
<p>The are some differences with the matrix implementation: * there is a small aproximation in the initialization * the estimation of the dispersion parameter is forced to be genewise</p>
<p>The best way to use NewWave in this case is the same we see before but I will not run this code because it would be long.</p>
<div class="sourceCode" id="cb182"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># res4 &lt;- newWave(tenx, K=10, children=2,n_gene_par = 100, n_cell_par = 100, commondispersion = FALSE)</span></code></pre></div>
<p>The following figures explain the performance of NewWave in this dataset with differents approach:</p>
<p><img src="time_newWave.png" width="50%"></p>
<p><img src="memory_newWave.png" width="50%"></p>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session Information</h1>
<div class="sourceCode" id="cb183"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R Under development (unstable) (2020-11-02 r79396)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] TENxBrainData_1.11.0        HDF5Array_1.19.0           
##  [3] rhdf5_2.35.0                DelayedArray_0.17.5        
##  [5] NewWave_1.1.1               SharedObject_1.5.5         
##  [7] mclust_5.4.7                ggplot2_3.3.2              
##  [9] Rtsne_0.15                  irlba_2.3.3                
## [11] Matrix_1.2-18               splatter_1.15.1            
## [13] SingleCellExperiment_1.13.3 SummarizedExperiment_1.21.1
## [15] Biobase_2.51.0              GenomicRanges_1.43.1       
## [17] GenomeInfoDb_1.27.3         IRanges_2.25.5             
## [19] S4Vectors_0.29.6            BiocGenerics_0.37.0        
## [21] MatrixGenerics_1.3.0        matrixStats_0.57.0         
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                  fs_1.5.0                     
##  [3] bit64_4.0.5                   filelock_1.0.2               
##  [5] httr_1.4.2                    rprojroot_2.0.2              
##  [7] tools_4.1.0                   backports_1.2.1              
##  [9] R6_2.5.0                      DBI_1.1.0                    
## [11] colorspace_2.0-0              rhdf5filters_1.3.3           
## [13] withr_2.3.0                   tidyselect_1.1.0             
## [15] bit_4.0.4                     curl_4.3                     
## [17] compiler_4.1.0                textshaping_0.2.1            
## [19] desc_1.2.0                    labeling_0.4.2               
## [21] scales_1.1.1                  checkmate_2.0.0              
## [23] rappdirs_0.3.1                pkgdown_1.6.1                
## [25] systemfonts_0.3.2             stringr_1.4.0                
## [27] digest_0.6.27                 rmarkdown_2.6                
## [29] XVector_0.31.1                pkgconfig_2.0.3              
## [31] htmltools_0.5.0               fastmap_1.0.1                
## [33] dbplyr_2.0.0                  rlang_0.4.9                  
## [35] RSQLite_2.2.1                 shiny_1.5.0                  
## [37] farver_2.0.3                  generics_0.1.0               
## [39] BiocParallel_1.25.1           dplyr_1.0.2                  
## [41] RCurl_1.98-1.2                magrittr_2.0.1               
## [43] BiocSingular_1.7.0            GenomeInfoDbData_1.2.4       
## [45] Rcpp_1.0.5                    munsell_0.5.0                
## [47] Rhdf5lib_1.13.0               lifecycle_0.2.0              
## [49] stringi_1.5.3                 yaml_2.2.1                   
## [51] zlibbioc_1.37.0               BiocFileCache_1.15.1         
## [53] AnnotationHub_2.23.0          grid_4.1.0                   
## [55] blob_1.2.1                    promises_1.1.1               
## [57] ExperimentHub_1.17.0          crayon_1.3.4                 
## [59] lattice_0.20-41               beachmat_2.7.3               
## [61] locfit_1.5-9.4                knitr_1.30                   
## [63] pillar_1.4.7                  glue_1.4.2                   
## [65] BiocVersion_3.13.1            evaluate_0.14                
## [67] BiocManager_1.30.10           httpuv_1.5.4                 
## [69] vctrs_0.3.5                   gtable_0.3.0                 
## [71] purrr_0.3.4                   assertthat_0.2.1             
## [73] xfun_0.19                     mime_0.9                     
## [75] rsvd_1.0.3                    xtable_1.8-4                 
## [77] later_1.1.0.1                 ragg_0.4.0                   
## [79] tibble_3.0.4                  AnnotationDbi_1.53.0         
## [81] memoise_1.1.0                 interactiveDisplayBase_1.29.0
## [83] ellipsis_0.3.1</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Federico Agostinis.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
