<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>EuroBioc2020 NewWave workshop • SurfingNewWave</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="EuroBioc2020 NewWave workshop">
<meta property="og:description" content="SurfingNewWave">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SurfingNewWave</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette.html">EuroBioc2020 NewWave workshop</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/fedeago/SurfingNewWave">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="vignette_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="vignette_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>EuroBioc2020 NewWave workshop</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/fedeago/SurfingNewWave/blob/master/vignettes/vignette.Rmd"><code>vignettes/vignette.Rmd</code></a></small>
      <div class="hidden name"><code>vignette.Rmd</code></div>

    </div>

    
    
<div id="introduction-to-newwave" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-to-newwave" class="anchor"></a>Introduction to NewWave</h1>
<p>EuroBioc 2020: 14-18 December</p>
<div id="workshop-description" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h2>
<p>The fast development of single cell sequencing technologies in the recent years has generated a gap between the throughput of the experiments and the capability of analizing the generated data. In this package, we implement mini-batch stochastic gradient descent and the possibility to work with HDF5 files. We decided to use a negative binomial model following the observation that droplet sequencing technologies do not induce zero inflation in the data. Thanks to these improvements and the possibility of massively parallelize the estimation process using PSOCK clusters, we are able to speed up the computation.</p>
<div id="r-bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<ul>
<li>SharedObject</li>
<li>DelayedArray</li>
<li>SingleCellExperiment</li>
<li>
</ul>
</div>
</div>
<div id="workshop-goals-and-objectives" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h2>
</div>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span>
  <span class="op">{</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Oshlack/splatter">splatter</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">irlba</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jkrijthe/Rtsne">Rtsne</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://mclust-org.github.io/mclust/">mclust</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">NewWave</span><span class="op">)</span><span class="op">}</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>NewWave is a new package that assumes a Negative Binomial distributions for dimensionality reduction and batch effect removal. In order to reduce the memory consumption it uses a PSOCK cluster combined with the R package SharedObject that allow to share a matrix between different cores without memory duplication. Thanks to that we can massively parallelize the estimation process with huge benefit in terms of time consumption. We can reduce even more the time consumption using some minibatch approaches on the different steps of the optimization.</p>
<p>I am going to show how to use NewWave with example data generated with Splatter.</p>
<p>Splatter is a R package designed to generate single cell transcriptomic data with a high level of customization.</p>
<p>In our implementation we want a batch effect to be corrected, the additionals batch parameters are setted in order to create more complex batch effect instead of defoult.</p>
<p>Think to the group as a cell type, we will use those group to test the resulted low dimensional representation throught a cluster analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/splatter/man/newParams.html">newSplatParams</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">N</span><span class="op">=</span><span class="fl">1000</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span>
<span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/splatter/man/splatSimulate.html">splatSimulateGroups</a></span><span class="op">(</span><span class="va">params</span>,batchCells<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">N</span><span class="op">/</span><span class="fl">2</span>,<span class="va">N</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span>,
                           group.prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0.1</span>,<span class="fl">10</span><span class="op">)</span>,
                           de.prob <span class="op">=</span> <span class="fl">0.2</span>,
                           batch.facLoc <span class="op">=</span> <span class="fl">2</span>,
                           batch.facScale <span class="op">=</span> <span class="fl">1</span>,
                           verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </code></pre></div>
<p>The resoult of splatter is a SingleCellExperiment with the generated data in the assays counts.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 10000 1000 </span>
<span class="co">#&gt; metadata(1): Params</span>
<span class="co">#&gt; assays(6): BatchCellMeans BaseCellMeans ... TrueCounts counts</span>
<span class="co">#&gt; rownames(10000): Gene1 Gene2 ... Gene9999 Gene10000</span>
<span class="co">#&gt; rowData names(16): Gene BaseGeneMean ... DEFacGroup9 DEFacGroup10</span>
<span class="co">#&gt; colnames(1000): Cell1 Cell2 ... Cell999 Cell1000</span>
<span class="co">#&gt; colData names(4): Cell Batch Group ExpLibSize</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; altExpNames(0):</span></code></pre></div>
<p>As you can see we have the number of cell that we choose(N=500) an 10 000 genes. For this workshop we will choose only the 1000 most variable genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hvg</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">hvg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span>
<span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">hvg</span>,decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span>,<span class="op">]</span></code></pre></div>
<p>In a SingleCellExperiment the colData slot contains the sample level metadata. In our situation there are two important variable, Batch and Group.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">data</span><span class="op">)</span>
<span class="co">#&gt; DataFrame with 1000 rows and 4 columns</span>
<span class="co">#&gt;                 Cell       Batch    Group ExpLibSize</span>
<span class="co">#&gt;          &lt;character&gt; &lt;character&gt; &lt;factor&gt;  &lt;numeric&gt;</span>
<span class="co">#&gt; Cell1          Cell1      Batch1   Group5    53416.6</span>
<span class="co">#&gt; Cell2          Cell2      Batch1   Group4    53788.6</span>
<span class="co">#&gt; Cell3          Cell3      Batch1   Group8    69984.2</span>
<span class="co">#&gt; Cell4          Cell4      Batch1   Group8    63404.5</span>
<span class="co">#&gt; Cell5          Cell5      Batch1   Group9    55837.9</span>
<span class="co">#&gt; ...              ...         ...      ...        ...</span>
<span class="co">#&gt; Cell996      Cell996      Batch2   Group3    57537.0</span>
<span class="co">#&gt; Cell997      Cell997      Batch2   Group1    43552.6</span>
<span class="co">#&gt; Cell998      Cell998      Batch2   Group4    68324.0</span>
<span class="co">#&gt; Cell999      Cell999      Batch2   Group3    63489.0</span>
<span class="co">#&gt; Cell1000    Cell1000      Batch2   Group9    64149.7</span></code></pre></div>
<p><strong>IMPORTANT:</strong> For batch effecr removal the batch variable must be a factor</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data</span><span class="op">$</span><span class="va">Batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">Batch</span><span class="op">)</span></code></pre></div>
<p>NewWave takes as input raw data, not normalized. So now we have all the object needed to start.</p>
<p>We can see the how the cells are distributed between group and batch.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/irlba/man/prcomp_irlba.html">prcomp_irlba</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span>,n<span class="op">=</span><span class="fl">10</span><span class="op">)</span>
<span class="va">plot_data</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html">Rtsne</a></span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">x</span><span class="op">)</span><span class="op">$</span><span class="va">Y</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot_data</span><span class="op">$</span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">Batch</span>
<span class="va">plot_data</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">Group</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">plot_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">X1</span>,y<span class="op">=</span><span class="va">X2</span>,col<span class="op">=</span><span class="va">group</span>, shape<span class="op">=</span><span class="va">batch</span><span class="op">)</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-10-1.png" width="700"> Just for a later comparison let’s see the performance of clustering on PCA low dimensional representation.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">x</span>, <span class="fl">10</span><span class="op">)</span>

<span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html">adjustedRandIndex</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">$</span><span class="va">cluster</span>, <span class="va">data</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.03949584</span></code></pre></div>
<p>There is a clear batch effect between the cells.</p>
<p>Let’s try to correct it.</p>
</div>
<div id="newwave" class="section level1">
<h1 class="hasAnchor">
<a href="#newwave" class="anchor"></a>NewWave</h1>
<p>I am going to show different implementation and the suggested way to use them with the given hardware.</p>
<p>Some basic advise:</p>
<ul>
<li>Verbose option has default FALSE, in this vignette I will change it for explanatory intentions, don’t do it with big dataset because it can sensibly slower the computation</li>
<li>There are no concern about the dimension of mini-batches, I always used the 10% of the observations</li>
</ul>
<div id="standard-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#standard-usage" class="anchor"></a>Standard usage</h2>
<p>I am going to present the stadard implementation of NewWave, it uses 1 cores, it estimates a dispersion parameters equal for all observation and uses all the observation in the estimation process.</p>
<p>The K parameter represents the number of dimension of the latent rapresentation.</p>
<p>The X parameter is used to indicates which variable represent the batch effect, in the same way can be inserted other cell-related variable and if you need some gene related variable those can be inserted in V.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NewWave/man/newWave.html">newWave</a></span><span class="op">(</span><span class="va">data</span>, X <span class="op">=</span> <span class="st">"~Batch"</span>, K<span class="op">=</span><span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Time of setup</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.009   0.004   0.331 </span>
<span class="co">#&gt; Time of initialization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.063   0.004   0.995</span>
<span class="co">#&gt; Iteration 1</span>
<span class="co">#&gt; penalized log-likelihood = -2065512.54139726</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.132   0.057   1.162</span>
<span class="co">#&gt; after optimize dispersion = -1777332.75031906</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.000   0.000  11.992</span>
<span class="co">#&gt; after right optimization= -1729847.99936925</span>
<span class="co">#&gt; after orthogonalization = -1729844.4376134</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.027   0.012  11.859</span>
<span class="co">#&gt; after left optimization= -1723772.95382354</span>
<span class="co">#&gt; after orthogonalization = -1723772.60552776</span>
<span class="co">#&gt; Iteration 2</span>
<span class="co">#&gt; penalized log-likelihood = -1723772.60552776</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.257   0.072   1.210</span>
<span class="co">#&gt; after optimize dispersion = -1722794.58131023</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.001   0.000  11.907</span>
<span class="co">#&gt; after right optimization= -1722283.59437893</span>
<span class="co">#&gt; after orthogonalization = -1722283.50293497</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;    0.02    0.02   12.60</span>
<span class="co">#&gt; after left optimization= -1722075.24146541</span>
<span class="co">#&gt; after orthogonalization = -1722075.17358539</span>
<span class="co">#&gt; Iteration 3</span>
<span class="co">#&gt; penalized log-likelihood = -1722075.17358539</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.221   0.069   1.170</span>
<span class="co">#&gt; after optimize dispersion = -1722072.41346738</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.001   0.000  10.730</span>
<span class="co">#&gt; after right optimization= -1721952.74993397</span>
<span class="co">#&gt; after orthogonalization = -1721952.69517518</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.005   0.032  12.268</span>
<span class="co">#&gt; after left optimization= -1721870.70571689</span>
<span class="co">#&gt; after orthogonalization = -1721870.68328032</span>
<span class="co">#&gt; Iteration 4</span>
<span class="co">#&gt; penalized log-likelihood = -1721870.68328032</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.233   0.060   1.174</span>
<span class="co">#&gt; after optimize dispersion = -1721870.58928044</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.001   0.000   9.809</span>
<span class="co">#&gt; after right optimization= -1721806.24431404</span>
<span class="co">#&gt; after orthogonalization = -1721806.1939903</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.031   0.008  11.702</span>
<span class="co">#&gt; after left optimization= -1721756.96418646</span>
<span class="co">#&gt; after orthogonalization = -1721756.9473855</span></code></pre></div>
<p>In order to make it faster you can increase the number of cores using “children” parameter:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NewWave/man/newWave.html">newWave</a></span><span class="op">(</span><span class="va">data</span>,X <span class="op">=</span> <span class="st">"~Batch"</span>, K<span class="op">=</span><span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>, children<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; Time of setup</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.010   0.004   0.318 </span>
<span class="co">#&gt; Time of initialization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.048   0.008   0.788</span>
<span class="co">#&gt; Iteration 1</span>
<span class="co">#&gt; penalized log-likelihood = -2065512.54139846</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.123   0.028   1.114</span>
<span class="co">#&gt; after optimize dispersion = -1777332.7503252</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.000   0.000  12.266</span>
<span class="co">#&gt; after right optimization= -1729847.99935577</span>
<span class="co">#&gt; after orthogonalization = -1729844.43759999</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.007   0.012   5.972</span>
<span class="co">#&gt; after left optimization= -1723772.953446</span>
<span class="co">#&gt; after orthogonalization = -1723772.60514951</span>
<span class="co">#&gt; Iteration 2</span>
<span class="co">#&gt; penalized log-likelihood = -1723772.60514951</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.223   0.084   1.180</span>
<span class="co">#&gt; after optimize dispersion = -1722794.58187868</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.001   0.000  13.286</span>
<span class="co">#&gt; after right optimization= -1722283.59422189</span>
<span class="co">#&gt; after orthogonalization = -1722283.5027724</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.017   0.012   6.238</span>
<span class="co">#&gt; after left optimization= -1722075.24928323</span>
<span class="co">#&gt; after orthogonalization = -1722075.18139377</span>
<span class="co">#&gt; Iteration 3</span>
<span class="co">#&gt; penalized log-likelihood = -1722075.18139377</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.232   0.076   1.188</span>
<span class="co">#&gt; after optimize dispersion = -1722072.42163634</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.001   0.001  10.734</span>
<span class="co">#&gt; after right optimization= -1721952.51857087</span>
<span class="co">#&gt; after orthogonalization = -1721952.46385013</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.016   0.024   6.303</span>
<span class="co">#&gt; after left optimization= -1721870.43370873</span>
<span class="co">#&gt; after orthogonalization = -1721870.41107056</span>
<span class="co">#&gt; Iteration 4</span>
<span class="co">#&gt; penalized log-likelihood = -1721870.41107056</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.256   0.048   1.184</span>
<span class="co">#&gt; after optimize dispersion = -1721870.31681561</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.000   0.000  10.087</span>
<span class="co">#&gt; after right optimization= -1721806.27238247</span>
<span class="co">#&gt; after orthogonalization = -1721806.2217882</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.021   0.016   5.911</span>
<span class="co">#&gt; after left optimization= -1721756.8884599</span>
<span class="co">#&gt; after orthogonalization = -1721756.87158308</span></code></pre></div>
</div>
<div id="commonwise-dispersion-and-minibatch-approaches" class="section level2">
<h2 class="hasAnchor">
<a href="#commonwise-dispersion-and-minibatch-approaches" class="anchor"></a>Commonwise dispersion and minibatch approaches</h2>
<p>If you do not have an high number of cores to run newWave this is the fastest way to run.</p>
<p>The optimization process is done by three process itereated until convercence.</p>
<ul>
<li>Optimization of the dispersion parameters</li>
<li>Optimization of the gene related parameters</li>
<li>Optimization of the cell related parameters</li>
</ul>
<p>Each of these three steps can be accelerated using mini batch, the number of observation is settled with these parameters:</p>
<ul>
<li>n_gene_disp : Number of genes to use in the dispersion optimization</li>
<li>n_cell_par : Number of cells to use in the cells related parameters optimization</li>
<li>n_gene_par : Number of genes to use in the genes related parameters optimization</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NewWave/man/newWave.html">newWave</a></span><span class="op">(</span><span class="va">data</span>,X <span class="op">=</span> <span class="st">"~Batch"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>,K<span class="op">=</span><span class="fl">10</span>, children<span class="op">=</span><span class="fl">2</span>,
                n_gene_disp <span class="op">=</span> <span class="fl">100</span>, n_gene_par <span class="op">=</span> <span class="fl">100</span>, n_cell_par <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="co">#&gt; Time of setup</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.010   0.004   0.320 </span>
<span class="co">#&gt; Time of initialization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.056   0.000   0.732</span>
<span class="co">#&gt; Iteration 1</span>
<span class="co">#&gt; penalized log-likelihood = -2065512.54139783</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.121   0.024   1.107</span>
<span class="co">#&gt; after optimize dispersion = -1777332.75031023</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.001   0.000  13.254</span>
<span class="co">#&gt; after right optimization= -1729847.99934542</span>
<span class="co">#&gt; after orthogonalization = -1729844.43758966</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.028   0.012   6.103</span>
<span class="co">#&gt; after left optimization= -1723772.95343943</span>
<span class="co">#&gt; after orthogonalization = -1723772.60514294</span>
<span class="co">#&gt; Iteration 2</span>
<span class="co">#&gt; penalized log-likelihood = -1723772.60514294</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.438   0.064   0.383</span>
<span class="co">#&gt; after optimize dispersion = -1722908.07782344</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.001   0.000   2.677</span>
<span class="co">#&gt; after right optimization= -1722790.90862571</span>
<span class="co">#&gt; after orthogonalization = -1722790.89469987</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.026   0.012   0.676</span>
<span class="co">#&gt; after left optimization= -1722784.94246851</span>
<span class="co">#&gt; after orthogonalization = -1722784.94212197</span>
<span class="co">#&gt; Iteration 3</span>
<span class="co">#&gt; penalized log-likelihood = -1722784.94212197</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.454   0.049   0.383</span>
<span class="co">#&gt; after optimize dispersion = -1722750.73885448</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.000   0.000   2.423</span>
<span class="co">#&gt; after right optimization= -1722674.30694858</span>
<span class="co">#&gt; after orthogonalization = -1722674.30108943</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.014   0.004   0.653</span>
<span class="co">#&gt; after left optimization= -1722665.01451073</span>
<span class="co">#&gt; after orthogonalization = -1722665.01367221</span></code></pre></div>
</div>
<div id="genewise-dispersion-mini-batch" class="section level2">
<h2 class="hasAnchor">
<a href="#genewise-dispersion-mini-batch" class="anchor"></a>Genewise dispersion mini-batch</h2>
<p>If you have a lot of core disposable or you want to estimate a genewise dispersion parameter this is the fastes configuration:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NewWave/man/newWave.html">newWave</a></span><span class="op">(</span><span class="va">data</span>,X <span class="op">=</span> <span class="st">"~Batch"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>,K<span class="op">=</span><span class="fl">10</span>, children<span class="op">=</span><span class="fl">2</span>,
                n_gene_par <span class="op">=</span> <span class="fl">100</span>, n_cell_par <span class="op">=</span> <span class="fl">100</span>, commondispersion <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Time of setup</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.010   0.004   0.329 </span>
<span class="co">#&gt; Time of initialization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.054   0.003   0.753</span>
<span class="co">#&gt; Iteration 1</span>
<span class="co">#&gt; penalized log-likelihood = -2065512.54139777</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.122   0.020   1.111</span>
<span class="co">#&gt; after optimize dispersion = -1777332.75031636</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.001   0.000  12.013</span>
<span class="co">#&gt; after right optimization= -1729847.99934963</span>
<span class="co">#&gt; after orthogonalization = -1729844.43759386</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.016   0.020   6.048</span>
<span class="co">#&gt; after left optimization= -1723772.95344137</span>
<span class="co">#&gt; after orthogonalization = -1723772.60514487</span>
<span class="co">#&gt; Iteration 2</span>
<span class="co">#&gt; penalized log-likelihood = -1723772.60514487</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.109   0.089   0.946</span>
<span class="co">#&gt; after optimize dispersion = -1716752.02580949</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.000   0.000   2.838</span>
<span class="co">#&gt; after right optimization= -1716661.56542251</span>
<span class="co">#&gt; after orthogonalization = -1716661.55445373</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.006   0.007   0.723</span>
<span class="co">#&gt; after left optimization= -1716616.75395123</span>
<span class="co">#&gt; after orthogonalization = -1716616.75306738</span>
<span class="co">#&gt; Iteration 3</span>
<span class="co">#&gt; penalized log-likelihood = -1716616.75306738</span>
<span class="co">#&gt; Time of dispersion optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.152   0.064   0.473</span>
<span class="co">#&gt; after optimize dispersion = -1716616.99593022</span>
<span class="co">#&gt; Time of right optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.001   0.000   2.477</span>
<span class="co">#&gt; after right optimization= -1716557.60878237</span>
<span class="co">#&gt; after orthogonalization = -1716557.60418941</span>
<span class="co">#&gt; Time of left optimization</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.008   0.020   0.688</span>
<span class="co">#&gt; after left optimization= -1716511.96121269</span>
<span class="co">#&gt; after orthogonalization = -1716511.96057207</span></code></pre></div>
<p>NB: do not use n_gene_disp in this case, it will slower the computation.</p>
<p>Now I can use the latent dimension rapresentation for visualization purpose:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">latent</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>

<span class="va">tsne_latent</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html">Rtsne</a></span><span class="op">(</span><span class="va">latent</span><span class="op">)</span><span class="op">$</span><span class="va">Y</span><span class="op">)</span>
<span class="va">tsne_latent</span><span class="op">$</span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">Batch</span>
<span class="va">tsne_latent</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">Group</span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">tsne_latent</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">X1</span>,y<span class="op">=</span><span class="va">X2</span>,col<span class="op">=</span><span class="va">group</span>, shape<span class="op">=</span><span class="va">batch</span><span class="op">)</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>or for clustering:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span><span class="op">(</span><span class="va">latent</span>, <span class="fl">10</span><span class="op">)</span>

<span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html">adjustedRandIndex</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">$</span><span class="va">cluster</span>, <span class="va">data</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.8703356</span></code></pre></div>
</div>
</div>
<div id="newwave-on-delayedarray" class="section level1">
<h1 class="hasAnchor">
<a href="#newwave-on-delayedarray" class="anchor"></a>NewWave on DelayedArray</h1>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TENxBrainData</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: HDF5Array</span>
<span class="co">#&gt; Loading required package: DelayedArray</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'DelayedArray'</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     aperm, apply, rowsum, scale, sweep</span>
<span class="co">#&gt; Loading required package: rhdf5</span></code></pre></div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session Information</h1>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R Under development (unstable) (2020-12-04 r79554)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.1 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] TENxBrainData_1.11.0        HDF5Array_1.19.0           </span>
<span class="co">#&gt;  [3] rhdf5_2.35.0                DelayedArray_0.17.4        </span>
<span class="co">#&gt;  [5] NewWave_1.1.1               SharedObject_1.5.5         </span>
<span class="co">#&gt;  [7] mclust_5.4.7                ggplot2_3.3.2              </span>
<span class="co">#&gt;  [9] Rtsne_0.15                  irlba_2.3.3                </span>
<span class="co">#&gt; [11] Matrix_1.2-18               splatter_1.15.1            </span>
<span class="co">#&gt; [13] SingleCellExperiment_1.13.3 SummarizedExperiment_1.21.0</span>
<span class="co">#&gt; [15] Biobase_2.51.0              GenomicRanges_1.43.0       </span>
<span class="co">#&gt; [17] GenomeInfoDb_1.27.1         IRanges_2.25.4             </span>
<span class="co">#&gt; [19] S4Vectors_0.29.5            BiocGenerics_0.37.0        </span>
<span class="co">#&gt; [21] MatrixGenerics_1.3.0        matrixStats_0.57.0         </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] bitops_1.0-6                  fs_1.5.0                     </span>
<span class="co">#&gt;  [3] bit64_4.0.5                   filelock_1.0.2               </span>
<span class="co">#&gt;  [5] httr_1.4.2                    rprojroot_2.0.2              </span>
<span class="co">#&gt;  [7] tools_4.1.0                   backports_1.2.0              </span>
<span class="co">#&gt;  [9] R6_2.5.0                      DBI_1.1.0                    </span>
<span class="co">#&gt; [11] colorspace_2.0-0              rhdf5filters_1.3.2           </span>
<span class="co">#&gt; [13] withr_2.3.0                   tidyselect_1.1.0             </span>
<span class="co">#&gt; [15] bit_4.0.4                     curl_4.3                     </span>
<span class="co">#&gt; [17] compiler_4.1.0                textshaping_0.2.1            </span>
<span class="co">#&gt; [19] desc_1.2.0                    labeling_0.4.2               </span>
<span class="co">#&gt; [21] scales_1.1.1                  checkmate_2.0.0              </span>
<span class="co">#&gt; [23] rappdirs_0.3.1                pkgdown_1.6.1                </span>
<span class="co">#&gt; [25] systemfonts_0.3.2             stringr_1.4.0                </span>
<span class="co">#&gt; [27] digest_0.6.27                 rmarkdown_2.5                </span>
<span class="co">#&gt; [29] XVector_0.31.0                pkgconfig_2.0.3              </span>
<span class="co">#&gt; [31] htmltools_0.5.0               fastmap_1.0.1                </span>
<span class="co">#&gt; [33] dbplyr_2.0.0                  rlang_0.4.9                  </span>
<span class="co">#&gt; [35] RSQLite_2.2.1                 shiny_1.5.0                  </span>
<span class="co">#&gt; [37] farver_2.0.3                  generics_0.1.0               </span>
<span class="co">#&gt; [39] BiocParallel_1.25.1           dplyr_1.0.2                  </span>
<span class="co">#&gt; [41] RCurl_1.98-1.2                magrittr_2.0.1               </span>
<span class="co">#&gt; [43] BiocSingular_1.7.0            GenomeInfoDbData_1.2.4       </span>
<span class="co">#&gt; [45] Rcpp_1.0.5                    munsell_0.5.0                </span>
<span class="co">#&gt; [47] Rhdf5lib_1.13.0               lifecycle_0.2.0              </span>
<span class="co">#&gt; [49] stringi_1.5.3                 yaml_2.2.1                   </span>
<span class="co">#&gt; [51] zlibbioc_1.37.0               BiocFileCache_1.15.1         </span>
<span class="co">#&gt; [53] AnnotationHub_2.23.0          grid_4.1.0                   </span>
<span class="co">#&gt; [55] blob_1.2.1                    promises_1.1.1               </span>
<span class="co">#&gt; [57] ExperimentHub_1.17.0          crayon_1.3.4                 </span>
<span class="co">#&gt; [59] lattice_0.20-41               beachmat_2.7.2               </span>
<span class="co">#&gt; [61] locfit_1.5-9.4                knitr_1.30                   </span>
<span class="co">#&gt; [63] pillar_1.4.7                  glue_1.4.2                   </span>
<span class="co">#&gt; [65] BiocVersion_3.13.1            evaluate_0.14                </span>
<span class="co">#&gt; [67] BiocManager_1.30.10           httpuv_1.5.4                 </span>
<span class="co">#&gt; [69] vctrs_0.3.5                   gtable_0.3.0                 </span>
<span class="co">#&gt; [71] purrr_0.3.4                   assertthat_0.2.1             </span>
<span class="co">#&gt; [73] xfun_0.19                     mime_0.9                     </span>
<span class="co">#&gt; [75] rsvd_1.0.3                    xtable_1.8-4                 </span>
<span class="co">#&gt; [77] later_1.1.0.1                 ragg_0.4.0                   </span>
<span class="co">#&gt; [79] tibble_3.0.4                  AnnotationDbi_1.53.0         </span>
<span class="co">#&gt; [81] memoise_1.1.0                 interactiveDisplayBase_1.29.0</span>
<span class="co">#&gt; [83] ellipsis_0.3.1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Federico Agostinis.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
